src/ui/mod.rs

```diff
@@ -0,0 +1,283 @@
+use indicatif::{MultiProgress, ProgressBar, ProgressStyle};
+use indicatif_log_bridge::LogWrapper;
+use itertools::Itertools;
+use std::collections::HashMap;
+
+#[cfg(test)]
+pub(crate) mod mock;
+
+pub(crate) trait ProofUI {
+    fn new(num_proofs: u64) -> Self;
+    fn println(&self, line: &str) -> std::io::Result<()>;
+
+    fn start_proof(&mut self, proof_name: &str, num_proofsteps: u64);
+
+    fn finish_proof(&mut self, proof_name: &str);
+
+    fn start_proofstep(&mut self, proof_name: &str, proofstep_name: &str);
+
+    fn proofstep_is_reduction(&mut self, proof_name: &str, proofstep_name: &str);
+
+    fn proofstep_set_oracles(&mut self, proof_name: &str, proofstep_name: &str, num_oracles: u64);
+
+    fn finish_proofstep(&mut self, proof_name: &str, proofstep_name: &str);
+
+    fn start_oracle(
+        &mut self,
+        proof_name: &str,
+        proofstep_name: &str,
+        oracle_name: &str,
+        num_lemmata: u64,
+    );
+
+    fn finish_oracle(&mut self, proof_name: &str, proofstep_name: &str, oracle_name: &str);
+
+    fn start_lemma(
+        &mut self,
+        proof_name: &str,
+        proofstep_name: &str,
+        oracle_name: &str,
+        lemma_name: &str,
+    );
+
+    fn finish_lemma(
+        &mut self,
+        proof_name: &str,
+        proofstep_name: &str,
+        oracle_name: &str,
+        lemma_name: &str,
+    );
+}
+
+pub(crate) struct IndicatifProofUI {
+    main_progress: MultiProgress,
+    project_progress: Option<ProgressBar>,
+    seq_proof_progress: HashMap<String, ProgressBar>,
+    seq_proofstep_progress: HashMap<(String, String), ProgressBar>,
+    seq_oracle_progress: HashMap<(String, String, String), ProgressBar>,
+}
```

Comment:
Could we move this, the impl and the style module into a submodule indicatif.rs?